old/launcher/invoker.1.in
.\" $Id: invoker.1.in 2355 2006-02-17 12:56:49Z guillem $
.TH maemo\-invoker 1 2006-02-17 @VERSION@ "Maemo Launcher manual"
.SH NAME
maemo\-invoker \- invoke a program from the launcher
.SH SYNOPSIS
.B maemo\-invoker
.RI [ options "] " file " [" file-options... ]
.SH DESCRIPTION
\fBmaemo\-invoker\fP tells \fBmaemo\-launcher\fP what application to start.
It will invoke the file provided on the command line with its optional
arguments, otherwise it will invoke its own program name with the string
\fI.launch\fP appended and will pass the arguments as well.
So it can be symlinked in place of the real \fIapplication\fP, and this
one moved to \fIapplication.launch\fP.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-\-delay \fISECS\fP
After invoking the program sleep for \fISECS\fP seconds before exiting. If
\fISECS\fP is 0 \fBmaemo\-invoker\fP will wait until the program finishes,
returning its exit status. The default behaviour is to wait for the invoked
program to exit.
.TP
.B \-\-help
Show summary of options.
.TP
.B \-\-version
Show version of program.
.SH ENVIRONMENT
.TP
.B MAEMO_INVOKER_DELAY
Set the exit delay. The command line option overrides this.
.SH SEE ALSO
.BR maemo\-launcher (1).
.br
.SH AUTHOR
\fBmaemo\-invoker\fP was written by Guillem Jover <guillem.jover@nokia.com>.


old/ChangeLog
2008-12-19  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.32.

2008-12-19  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Add --enable-qt-booster.
	* launcher/Makefile.am (booster_qt_so_SOURCES): New variable.
	(booster_qt_so_CFLAGS): Likewise.
	(booster_qt_so_LDFLAGS): Likewise.
	(booster_qt_so_CPPFLAGS): Likewise.
	(booster_qt_so_LDADD): Likewise.
	[USE_QT_BOOSTER] (booster_PROGRAMS): Likewise.
	* launcher/booster_qt.c: New file.

2008-12-19  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Bump version to 0.32~.

2008-07-03  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.31.

2008-07-03  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/invoker.c (disable_self_core_dump): New function.
	(main): Call disable_self_core_dump before setting up the signal
	forwarder.

2008-07-03  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Bump version to 0.31~.

2008-05-29  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.30.

2008-05-28  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac (HILDON_DEPENDS): Replace libossohelp with hildon-help.

2008-05-12  Eero Tamminen  <eero.tamminen@nokia.com>

	* README: Typo fix and additional information for symbol versioning.

2008-05-12  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Bump version to 0.30~.

2008-05-05  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.29.

2008-05-05  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/launcher.c (kindergarten_destroy): Traverse the whole
	child list and act on childs with a non-zero pid.
	(store_state): Likewise.

2008-05-05  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/launcher.c (child_destroy): Reset pid at the beginning.

2008-05-05  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Bump version to 0.29~.

2008-04-27  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.28.

2008-04-27  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/launcher.c (rise_oom_defense): Change error to warning
	when the defender failed to execute.

2008-04-27  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/report.h (enum report_type): Add report_warning.
	(warning): New macro.
	* launcher/report.c (vreport): Handle report_warning.

2008-04-23  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/launcher.c (store_state): Call comm_msg_destroy.
	(load_state): Likewise.

2008-04-23  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Bump version to 0.28~.

2008-04-15  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.27.

2008-04-15  Guillem Jover  <guillem.jover@nokia.com>

	* README: Add some documentation about version scripts.

2008-04-15  Guillem Jover  <guillem.jover@nokia.com>

	* README: Add some documentation for make based build systems.

2008-04-15  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.h (comm_msg_set_magic): Rename to ...
	(comm_msg_put_magic): ... this. Fix all users.
	(comm_msg_pack_int): Rename to ...
	(comm_msg_put_int): ... this. Fix all users.
	(comm_msg_unpack_int): Rename to ...
	(comm_msg_get_int): ... this. Fix all users.
	(comm_msg_pack_str): Rename to ...
	(comm_msg_put_str): ... this. Fix all users.
	(comm_msg_unpack_str): Rename to ...
	(comm_msg_get_str): ... this. Fix all users.

2008-04-15  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.c (comm_msg_unpack_mem): Remove function.
	(comm_msg_put_u32): New function.
	(comm_msg_get_u32): Likewise.
	(comm_msg_set_magic): Use comm_msg_put_u32 instead of direct buffer
	access.
	(comm_msg_pack_int): Likewise.
	(comm_msg_pack_str): Likewise.
	(comm_msg_get_magic): Use comm_msg_get_u32 instead of direct buffer
	access.
	(comm_msg_unpack_int): Likewise. Perform boundary checks.
	(comm_msg_unpack_str): Likewise. Get the string through the current
	buffer read address.

2008-03-20  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.h (comm_msg_set_magic): New prototype.
	(comm_msg_get_magic): Likewise.
	* launcher/comm_msg.c (comm_msg_set_magic): New function.
	(comm_msg_get_magic): Likewise.
	* launcher/test_msg.c (main): Set and get the magic and check the
	value being the same.
	* launcher/launcher.c (LAUNCHER_STATE_SIG): Change from "MLSF0.0"
	to 0x30534c4d.
	(store_state): Use comm_msg_set_magic instead of comm_msg_pack_str.
	(load_state): Use comm_msg_get_magic instead of comm_msg_unpack_str.

2008-03-18  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.h (comm_msg_print): New prototype.
	* launcher/comm_msg.c (struct comm_msg): Change buf member from char *
	to void *.
	(comm_msg_print): New function.
	(comm_msg_send): Use comm_msg_print to print msg instead of debug.
	(comm_msg_recv): Likewise.

2008-03-14  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.c (comm_msg_send): Use msg->used in sizeof instead
	of msg->size.

2008-03-07  Guillem Jover  <guillem.jover@nokia.com>

	* README: Fix a typo and add a missing word.

2008-03-07  Guillem Jover  <guillem.jover@nokia.com>

	* client/client.c (cleanup): New function.
	(main): Install cleanup as an atexit function.
	* launcher/launcher.c (launch_process): Call exit instead of _exit.
	* launcher/summoner.c (summon_process): Likewise.
	(main): Return 0 at the end of the function instead of _exit.

2008-03-07  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac (CLIENT): Use only the minimum dependencies,
	GTK_DEPENDS and hildon-1.

2008-02-22  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.c (comm_msg_pack_mem): Move code to ...
	(comm_msg_pack_str): ... here.
	(comm_msg_pack_int): Rewrite specialized version without taking into
	account unaligned memory to improve performance.

2008-02-22  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/Makefile.am (check_PROGRAMS): Add 'test-msg-perf'.
	(test_msg_perf_SOURCES): New variable.
	(test_msg_perf_CPPFLAGS): Likewise.

src/invoker/invoker.c
/*
 * invoker.c
 *
 * This file is part of applauncherd
 *
 * Copyright (C) 2010 Nokia Corporation. All rights reserved.
 *
 * This software, including documentation, is protected by copyright
 * controlled by Nokia Corporation. All rights are reserved.
 * Copying, including reproducing, storing, adapting or translating,
 * any or all of this material requires the prior written consent of
 * Nokia Corporation. This material also contains confidential
 * information which may not be disclosed to others without the prior
 * written consent of Nokia.
 */

#define _GNU_SOURCE

#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <sys/socket.h>
#include <bits/socket.h>
#include <sys/un.h>
#include <sys/uio.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <unistd.h>
#include <errno.h>
#include <sys/wait.h>

#include "report.h"
#include "invokelib.h"
#include "search.h"

static const int DEFAULT_DELAY = 0;

enum APP_TYPE { DUI_APP, QT_APP, UNKNOWN_APP };

extern char ** environ;


static bool invoke_recv_ack(int fd)
{
    uint32_t action;

    /* Receive ACK. */
    invoke_recv_msg(fd, &action);

    if (action != INVOKER_MSG_ACK)
    {
        die(1, "receiving wrong ack (%08x)\n", action);
    }
    else
        return true;
}

static int invoker_init(enum APP_TYPE app_type)
{
    int fd;
    struct sockaddr_un sun;

    fd = socket(PF_UNIX, SOCK_STREAM, 0);
    if (fd < 0)
        die(1, "opening invoker socket\n");

    sun.sun_family = AF_UNIX;  //AF_FILE;

    if(app_type ==  DUI_APP)
        strcpy(sun.sun_path, INVOKER_DUI_SOCK);
    else if (app_type ==  QT_APP)
        strcpy(sun.sun_path, INVOKER_QT_SOCK);
    else
        die(1, "unknown type of application: %d \n", app_type);

    if (connect(fd, (struct sockaddr *)&sun, sizeof(sun)) < 0)
        die(1, "connecting to the launcher\n");

    return fd;
}

static bool invoker_send_magic(int fd, int options)
{
    /* Send magic. */
    invoke_send_msg(fd, INVOKER_MSG_MAGIC | INVOKER_MSG_MAGIC_VERSION | options);
    invoke_recv_ack(fd);

    return true;
}

static bool invoker_send_name(int fd, char *name)
{
    /* Send action. */
    invoke_send_msg(fd, INVOKER_MSG_NAME);
    invoke_send_str(fd, name);
    invoke_recv_ack(fd);

    return true;
}

static bool invoker_send_exec(int fd, char *exec)
{
    /* Send action. */
    invoke_send_msg(fd, INVOKER_MSG_EXEC);
    invoke_send_str(fd, exec);
    invoke_recv_ack(fd);

    return true;
}

static bool invoker_send_args(int fd, int argc, char **argv)
{
    int i;

    /* Send action. */
    invoke_send_msg(fd, INVOKER_MSG_ARGS);
    invoke_send_msg(fd, argc);
    for (i = 0; i < argc; i++)
    {
        debug("param %d %s \n", i, argv[i]);
        invoke_send_str(fd, argv[i]);
    }
    invoke_recv_ack(fd);

    return true;
}

static bool invoker_send_env(int fd)
{
    int i, n_vars;

    /* Count the amount of environment variables. */
    for (n_vars = 0; environ[n_vars] != NULL; n_vars++) ;

    /* Send action. */
    invoke_send_msg(fd, INVOKER_MSG_ENV);
    invoke_send_msg(fd, n_vars);
    for (i = 0; i < n_vars; i++)
        invoke_send_str(fd, environ[i]);

    return true;
}

static bool invoker_send_io(int fd)
{
    struct msghdr msg;
    struct cmsghdr *cmsg;
    int io[3] = { 0, 1, 2 };
    char buf[CMSG_SPACE(sizeof(io))];
    struct iovec iov;
    int dummy;

    memset(&msg, 0, sizeof(struct msghdr));

    iov.iov_base = &dummy;
    iov.iov_len = 1;

    msg.msg_iov = &iov;
    msg.msg_iovlen = 1;
    msg.msg_control = buf;
    msg.msg_controllen = sizeof(buf);

    cmsg = CMSG_FIRSTHDR(&msg);
    cmsg->cmsg_len = CMSG_LEN(sizeof(io));
    cmsg->cmsg_level = SOL_SOCKET;
    cmsg->cmsg_type = SCM_RIGHTS;

    memcpy(CMSG_DATA(cmsg), io, sizeof(io));

    msg.msg_controllen = cmsg->cmsg_len;

    invoke_send_msg(fd, INVOKER_MSG_IO);
    if (sendmsg(fd, &msg, 0) < 0)
    {
        warning("sendmsg failed in invoker_send_io: %s /n", strerror(errno));
        return  false;
    }

    return true;
}

static bool invoker_send_end(int fd)
{
    /* Send action. */
    invoke_send_msg(fd, INVOKER_MSG_END);
    invoke_recv_ack(fd);

    return true;
}

static void version(void)
{
    printf("%s (%s) %s\n", PROG_NAME, PACKAGE, VERSION);
    exit(0);
}

static void usage(int status)
{
    printf("\nUsage: %s TYPE [options] file [file-options]\n"
           "Launch dui or qt application. \n"
           "Example %s --type=dui /usr/bin/helloworld \n"
           "TYPE: \n"
           "  dui               Launch DUI application. \n"
           "  qt                Launch Qt application.  \n"
           "Options:\n"
           "  --delay SECS        After invoking sleep for SECS seconds (default %d).\n"
           "  --version           Print program version.\n"
           "  --help              Print this help message.\n",
           PROG_NAME, PROG_NAME, DEFAULT_DELAY);

    exit(status);
}

int main(int argc, char *argv[])
{
    int i;
    int fd;
    int prog_argc = 0;
    char **prog_argv = NULL;
    char *prog_name = NULL;
    char *delay_str = NULL;
    int magic_options = 0;
    enum APP_TYPE app_type = UNKNOWN_APP;

    if (strstr(argv[0], PROG_NAME))
    {
        /* Check application type to start */
        if (argc < 2)
        {
            report(report_error, "application type is missing \n");
            usage(1);
        }
        else if (strcmp(argv[1], "--type=dui") == 0)
            app_type = DUI_APP;
        else if (strcmp(argv[1], "--type=qt") == 0)
            app_type = QT_APP;
        else if (strcmp(argv[1], "--version") == 0)
            version();
        else if (strcmp(argv[1], "--help") == 0)
            usage(0);
        else
        {
            report(report_error, "unknown type of application %s \n", argv[1]);
            usage(1);
        }

        /* Parse invoker options and search binary to launch */
        if (argc < 3)
        {
            report(report_error, "application name is missing \n");
            usage(1);
        }
        for (i = 2; i < argc; ++i)
        {
            if (strcmp(argv[i], "--delay") == 0)
            {
                if (argv[++i])
                    delay_str = argv[i];

            }
            else if (strcmp(argv[i], "--version") == 0)
                continue;
            else if (strcmp(argv[i], "--help") == 0)
                continue;
            else if (strncmp(argv[i], "--", 2) == 0)
            {
                report(report_error, "unknown parameter %s \n", argv[i]);
                usage(1);
            }
            else
            {
                prog_name = search_program(argv[i]);
                if (!prog_name)
                {
                    report(report_error, "can't find application to invoke\n");
                    usage(0);
                }

                char *period = strstr(argv[i], ".launch");
                if (period)
                    *period = '\0';

                prog_argc = argc - i;
                prog_argv = &argv[i];
                break;
            }
        }
    }
    else
    {
        /* Called with a different name, old way of using invoker */
        die(1, "wrong way of using, don't use symlinks");
    }

    info("invoking execution: '%s'\n", prog_name);

    fd = invoker_init(app_type);

    invoker_send_magic(fd, magic_options);
    invoker_send_name(fd, prog_argv[0]);
    invoker_send_exec(fd, prog_name);
    invoker_send_args(fd, prog_argc, prog_argv);
    invoker_send_io(fd);
    invoker_send_env(fd);
    invoker_send_end(fd);

    if (prog_name)
        free(prog_name);

    close(fd);

    return 0;
}

